### DISCUSSION TITLE
Ensure all visuals have alt text for accessibility

### DISCUSSION THREAD

**Discussion by cbaragao** [NatVanG/PBIXInspector#10] [Ideas]
In the json it is listed under general if it is there:

\"general\":[{\"properties\":{\"altText\":{\"expr\":{\"Literal\":{\"Value\":\"'TEST alt text'\"}}}}}]

**Comment by NatVanG** [comment]
I think the rule below addresses this requirement, I'll add it to the project's base rule files. Do you think that the rule should apply to each visual, only specific types of visuals or a percentage threshold over the whole report? We may need to look at a custom operator to trim whitespace characters so any altText with just whitespace is caught too.
```
{
  "name": "Ensure alternativeText has been defined for all visuals",
  "description": "Alt-text is required for screenreaders",
  "disabled": false,
  "logType": "warning",
  "forEachPath": "$.sections[*]",
  "forEachPathName": "$.name",
  "forEachPathDisplayName": "$.displayName",
  "path": "$.visualContainers[*].config",
  "pathErrorWhenNoMatch": false,
  "test": [
    {
      "map": [
        {
          "filter": [
            {
              "var": "visualsConfigArray"
            },
            {
              "none": [
                {
                  "var": "singleVisual.vcObjects.general"
                },
                {
                  "or": [
                    {
                      "!!": [ { "var": "properties.altText.expr.Aggregation" } ]
                    },
                    {
                      "!=": [ { "var": "properties.altText.expr.Literal.Value" }, "''" ]
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "var": "name"
        }
      ]
    },
    {
      "visualsConfigArray": "."
    },
    []
  ]
}
```

**Reply by cbaragao** [reply]
I think this should be a warning at least to let them know, perhaps if more than 80% has no alt text this could be called out?

**Reply by cbaragao** [reply]
This may be helpful, it is what the Lighthouse tool uses as guidance: https://dequeuniversity.com/rules/axe/4.7/image-alt
